lib/submit-request.js

```diff
@@ -40,9 +41,51 @@ function SubmitRequest(backend, agent, index, id, op, options) {
   this.snapshot = null;
   this.ops = [];
   this.channels = null;
+  this._fixupOps = [];
 }
 module.exports = SubmitRequest;
 
+SubmitRequest.prototype.$fixup = function(op) {
+  if (this.action !== this.backend.MIDDLEWARE_ACTIONS.apply) {
+    return new ShareDBError(
+      ERROR_CODE.ERR_FIXUP_IS_ONLY_VALID_ON_APPLY,
+      'fixup can only be called during the apply middleware'
+    );
+  }
+
+  if (this.op.del) {
+    return new ShareDBError(
+      ERROR_CODE.ERR_CANNOT_FIXUP_DELETION,
+      'fixup cannot be applied on deletion ops'
+    );
+  }
+
+  var typeId = this.op.create ? this.op.create.type : this.snapshot.type;
+  var type = types.map[typeId];
+  if (typeof type.compose !== 'function') {
+    return new ShareDBError(
+      ERROR_CODE.ERR_TYPE_DOES_NOT_SUPPORT_COMPOSE,
+      typeId + ' does not support compose'
+    );
+  }
+
+  try {
+    if (this.op.create) this.op.create.data = type.apply(this.op.create.data, op);
+    else this.op.op = type.compose(this.op.op, op);
+  } catch (error) {
+    return error;
```

Comment:
Let's throw instead of return errors, because errors will be more visible and force consumers to handle them.